certora/specs/YieldStreams.spec

```diff
@@ -0,0 +1,800 @@
+ */
+
+rule integrity_of_topUp(uint256 _streamId, uint256 _shares) {
+    env e;
+    address streamer = e.msg.sender;
+    address receiver = streamIdToReceiver(_streamId);
+    uint256 principal;
+
+    // Preconditions
+    require _shares > 0;
+    require ownerOf(_streamId) == streamer;
+    require receiver != 0;
+
+    // Capture the initial state
+    uint256 initialTotalShares = receiverTotalShares(receiver);
+    uint256 initialTotalPrincipal = receiverTotalPrincipal(receiver);
+    uint256 initialPrincipal = receiverPrincipal(receiver, _streamId);
+
+    // Call the `topUp` function
+    principal = topUp(e, _streamId, _shares);
+
+    // Postconditions
+    assert to_mathint(receiverTotalShares(receiver)) == initialTotalShares + _shares;
+    assert to_mathint(receiverTotalPrincipal(receiver)) == initialTotalPrincipal + principal;
+    assert to_mathint(receiverPrincipal(receiver, _streamId)) == initialPrincipal + principal;
+}
+
+/**
+ * @Rule Integrity property for the `topUpPermit` function
+ * @Category High
+ * @Description Ensures that the `topUpPermit` function correctly adds additional shares to an existing yield stream and updates the contract state.
+ */
+
+rule integrity_of_topUpPermit(uint256 _streamId, uint256 _shares, uint256 deadline, uint8 v, bytes32 r, bytes32 s) {
+    env e;
+    address streamer = e.msg.sender;
+    address receiver = streamIdToReceiver(_streamId);
+    uint256 principal;
+
+    // Preconditions
+    require _shares > 0;
+    require ownerOf(_streamId) == streamer;
+    require receiver != 0;
+
+    // Capture the initial state
+    uint256 initialTotalShares = receiverTotalShares(receiver);
+    uint256 initialTotalPrincipal = receiverTotalPrincipal(receiver);
+    uint256 initialPrincipal = receiverPrincipal(receiver, _streamId);
+    //uint256 initialStreamerShareBalance = vault.balanceOf(streamer);
+
+    // Call the `topUp` function
+    principal = topUpUsingPermit(e, _streamId, _shares, deadline, v, r, s);
+
+    // Postconditions
+    assert to_mathint(receiverTotalShares(receiver)) == initialTotalShares + _shares;
+    assert to_mathint(receiverTotalPrincipal(receiver)) == initialTotalPrincipal + principal;
+    assert to_mathint(receiverPrincipal(receiver, _streamId)) == initialPrincipal + principal;
+}
+
+
+/**
+ * @Rule Integrity property for the `depositAndTopUp` function
+ * @Category High
+ * @Description Ensures that the `depositAndTopUp` function adds the specified principal amount to the existing yield stream, deposit the principal, and update the contract state accordingly, if the caller is the owner of the stream.
+ */
+ 
+rule integrity_of_depositAndTopUp(uint256 _streamId, uint256 _principal) {
+    env e;
+    address streamer = e.msg.sender;
+    address receiver = streamIdToReceiver(_streamId);
+    uint256 shares;
```

Comment:
it will be easier to read if moving the declaration to the assignment, i.e.,
`uint256 shares = depositAndTopUp(e, _streamId, _principal);`

Reply:
Updated